classes/file_search.php

```diff
@@ -0,0 +1,423 @@
+            }
+            $zip->close();
+        }
+        // Todo: handle exception if zip file cannot be openned.
+        unlink($tmpfile);
+        return $matchcount;
+    }
+
+
+    /**
+     * Search the file, looking for the regular expression.
+     * Report the matches into the stream.
+     *
+     * @param object $filerecord  A row from mdl_files table, indicating the file to be searched.
+     * @param string $pattern A regular expression to search for.
+     * @param resource $stream  The open csv file to receive the matches.
+     * @return int $matchcount The number of matches found.
+     */
+    public static function search_file(object $filerecord, string $pattern, $stream): int {
+        static $fs = null;
+        if (empty($fs)) {
+            $fs = get_file_storage();
+        }
+        $file = $fs->get_file(
+            $filerecord->contextid,
+            $filerecord->component,
+            $filerecord->filearea,
+            $filerecord->itemid,
+            $filerecord->filepath,
+            $filerecord->filename
+        );
+
+        $csv = [
+            self::CSV_CONTEXTID => $filerecord->contextid,
+            self::CSV_COMPONENT => $filerecord->component,
+            self::CSV_FILEAREA  => $filerecord->filearea,
+            self::CSV_ITEMID    => $filerecord->itemid,
+            self::CSV_FILEPATH  => $filerecord->filepath,
+            self::CSV_FILENAME  => $filerecord->filename,
+            self::CSV_MIMETYPE  => $filerecord->mimetype,
+
+        ];
+        switch ($filerecord->mimetype) {
+            case 'application/zip.h5p':
+                $csv[self::CSV_STRATEGY] = 'zip';
+                $matchcount = self::unzip_content($csv, $file->get_content(), $pattern, $stream);
+                break;
+            default:
+                $csv[self::CSV_STRATEGY] = 'plain';
+                $csv[self::CSV_INTERNAL] = '';
+                $matchcount = self::grep_content($csv, $file->get_content(), $pattern, $stream);
+                break;
+        }
+            return $matchcount;
+    }
+
+        /**
+         * Make a where clause to implement the filtering criteria.
+         *
+         * @param string $components Comma-seperated  component:area pairs.
+         * @param string $skipcomponents Comma-separated components to be omitted.
+         * @param string $skipareas Comma-separated areas to be omitted.
+         * @param string $mimetypes Comma-separated mimetypes to be searched.
+         * @param string $skipmimetypes Comma-separated mimetypes to be omitted.
+         * @param string $filenames Comma-separated filenames to be searched.
+         * @param string $skipfilenames Comma-separated filenames to be omitted.
+         * @return string $whereclause A where clause ready for SQL.
+         * @return array $params An array of parameters to go with the where clause.
+         */
+    public static function make_where_clause($components, $skipcomponents, $skipareas,
```

Comment:
Should we have some sensible defaults for when nothing is specified? Particularly with mimetypes

Reply:
Yes. Leaving out "binary" files is probably a good idea.  I would like to reserve this for a future commit.